### DISCUSSION TITLE
How to build menu with a search field

### DISCUSSION THREAD

**Discussion by deebov** [adobe/react-spectrum#5315] [Q&A] [answered]
What would be the best way to build a menu with a search field using using `react-aria-component`? 

You can see what I mean in the example of Notion. This is very similar to combobox but here the input field is kind of attached to the list and does not contain the current selected value but rather acts like a search field for menu items. I tried to just put a simple input alongside menu items but when you hover menu items the focus moves away from the input. But we'd like to have the focus on input even when you hover and navigate menu items with arrow keys. I also tried to patch the `MenuItem` component to pass down `shouldUseVirtualFocus` to `useMenuItem` but then the keydown events are not propagated to the menu element when the input is focused.

https://github.com/adobe/react-spectrum/assets/28673719/078145fd-de54-4ad3-b932-a02ad7b6b84e

**Comment by reidbarber** [comment]
I think you'll want to use a [Combobox](https://react-spectrum.adobe.com/react-aria/ComboBox.html), but just clear the input and selection once an item is selected.

**Reply by snowystinger** [reply]
You can also use a Popover and combine a textfield at the top with a list below it.

You cannot have a search inside a menu, that's not valid aria. See https://w3c.github.io/aria/#menu for allowed children

**Reply by reidbarber** [reply]
I guess the downside is that you'll have to tab to move focus down to the menu and back.

**Reply by deebov** [reply]
I think the biggest problem would be to keep the focus inside a text field while navigation thought the list since list items take away focus.

**Reply by deebov** [reply]
Any tips to overcome this challenge would be very helpful.

**Reply by deebov** [reply]
I actually tried building what I described using `useComboBox` and that turned out well. Except `useComboBox` does not support multi-selection which we do need.

**Reply by snowystinger** [reply]
The way combobox work is it keeps focus in the field and has a virtual (appearance only) focus which moves in the list
it might be useful for you to look at how our Mobile version of Combobox works https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/combobox/src/MobileComboBox.tsx since that sounds pretty close.

A combobox cannot support multi select, but if you're down at the hook level, you could swap out for multi select and rebuild useComboboxState on top of that. As long as it's a short list and not async, that should be fine in theory. However, I think you'll quickly run into some problems in terms of usability.

The reason we don't allow for it is because when the combobox is closed, there is no good way right now to show all the selected items and it isn't a great idea to hide them in an a popover which contains an incomplete list or very long list where users could miss what they have selected vs not. We also wouldn't know how to filter the list when it is reopened, would it be the first selected entry? second? last?

It may make more sense to have your textfield just be a filter with one value, the current filter, and then a list with multiple selection enabled below it.

**Reply by Pouet--** [reply]
@deebov Hello, I am trying to build the same behavior for a dropdown menu with a search input and a list with virtual focus. Do you have any example or advice with using the react-aria hooks ? Thank you in advance

**Reply by ozguruysal** [reply]
I need to build a Searchable Menu component too and drafted [this example](https://stackblitz.com/edit/vitejs-vite-1mfnjq?file=src%2FApp.tsx&terminal=dev) using ComboBox as suggested by @reidbarber. It's far from finished of course but the problem is when you click on the menu and hover on any item or type something to filter the options, I see this error being logged in the console:

```
TypeError: Cannot read properties of null (reading 'contains')
```

Is there a way to fix this? I'm happy to use the hooks too.

**Reply by snowystinger** [reply]
Two observations
1) It looks like aria-hideoutside mutation observer is responsible for the error message. We'd need to log out what the value is if it's not a node to see what's happening there.
2) When I delete the value in the field, the combobox doesn't repopulate for some reason, I'm not sure what is going on there either.

If you're open to using the hooks, then I suggest having a look at our implementation of MobileCombobox in React Spectrum. It doesn't use the new RAC collections yet, but the gist should be the same.

**Reply by ozguruysal** [reply]
Thank you @snowystinger 👍, I'll review that. 

Would be really nice to get that kind of composability at some point though, if technically feasible.

**Comment by selrond** [comment, marked as answer]
@deebov [Autocomplete](https://react-spectrum.adobe.com/react-aria/Autocomplete.html "Autocomplete – React Aria") component is now in beta, allowing for this exact case among others

**Reply by tleunen** [reply]
@selrond Does it also work with a GridList? The doc only mentions Menu or Listbox

**Reply by snowystinger** [reply]
It does not yet, but we are working on that support right now.

**Reply by selrond** [reply]
@tleunen there’s this line in the [latest (august) release post](https://react-spectrum.adobe.com/releases/2025-08-25.html):

> We have also advanced Autocomplete to RC, and added support for filtering GridList, Table, and TagGroup (virtual focus support is still to come)

Although I don’t see it documented just yet.
